pkg/apis/node/types.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RuntimeClass defines a class of container runtime supported in the cluster.
+// The RuntimeClass is used to determines which container runtime is used to run
+// all containers in a pod. RuntimeClasses are (currently) manually defined by a
+// user or cluster provisioner, and referenced in the PodSpec. The Kubelet is
+// responsible for resolving the RuntimeClassName reference before running the
+// pod.  For more details, see
+// https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md
+type RuntimeClass struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the RuntimeClass
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	Spec RuntimeClassSpec
+}
+
+// RuntimeClassSpec is a specification of a RuntimeClass. It contains parameters
+// that are required to describe the RuntimeClass to the Container Runtime
+// Interface (CRI) implementation, as well as any other components that need to
+// understand how the pod will be run. The RuntimeClassSpec should be treated as
+// immutable.
+type RuntimeClassSpec struct {
+	// RuntimeHandler specifies the underlying runtime and configuration that the
+	// CRI implementation will use to handle pods of this class. The possible
+	// values are specific to the node configuration & CRI implementation.  It is
+	// assumed that all handlers are available on every node, and handlers of the
+	// same name are equivalent on every node.
+	// For example, a handler called "runc" might specify that the runc OCI
+	// runtime (using native Linux containers) should be used to run the
+	// containers in a pod.
+	// If this is not specified (or empty), a default will be chosen by the
+	// implementation.
+	// The RuntimeHandler must conform to the DNS Subdomain (RFC 1123)
```

Comment:
Let's talk about contents.  What are the contents of this?  Where does a string that works here originate?  E.g. in storage class, the driver itself defines a string, which is expected to be a kubernetes qualified name (prefix.com/suffix) and is defined by the storage vendor.

If I were a cluster admin who wanted to use this, where would I get the value that I have to write here?  And is really anything allowed or should we require the same spec as storage(ish) ?

Reply:
See my reply in the other thread, here: https://github.com/kubernetes/kubernetes/pull/74433#discussion_r262181978

The value is essentially opaque to kubernetes, and it's up to the CRI implementation for how to use it.

So why not just use a completely unconstrained string?

1. Much harder to tighten validation than relax it.
2. Bounding the values makes it easier on the CRI impls to not worry as much about formatting, and makes it more portable to different config setups (e.g. the key is both valid TOML and yaml, although the dots complicate it a bit.
3. Prevents this from becoming an arbitrary configuration mechanism (I don't want this to become a command-line string for launching the runtime).

We could use an even more strict requirement (e.g. `[a-zA-Z][-_a-zA-Z0-9]+[a-zA-Z0-9]`), but RFC 1123 is convenient and consistent with other Kubernetes names.

Reply:
Conclusion from offline discussion: this field is opaque to kubernetes. It's up to the CRI implementation how to interpret it (e.g. if they want to namespace it). Action item is to clarify this in the doc comments.

Reply:
After IRL conversation: Document the hell out of this as "something the runtime understands, e.g. may be a key in a config file".   Does this need to be a subdomain, or is a DNS Label sufficient?  No dots...

Reply:
Let's go with the stricter DNS label.